internal/util/kubernetes/object/status/queue.go

```diff
@@ -46,57 +46,47 @@ func WithBufferSize(size int) QueueOption {
 func NewQueue(opts ...QueueOption) *Queue {
 	q := &Queue{
 		subscriptionBufferSize: DefaultBufferSize,
-		subscriptions:          make(map[string]chan event.GenericEvent),
+		subscriptions:          make(map[string][]chan event.GenericEvent),
 	}
 	for _, opt := range opts {
 		opt(q)
 	}
+
 	return q
 }
 
 // Publish emits a GenericEvent for the provided objects that indicates to
 // subscribers that the status of that object needs to be updated.
 // It's a no-op if there are no subscriptions for the object kind.
 func (q *Queue) Publish(obj client.Object) {
-	ch, ok := q.getSubscriptionForKind(obj.GetObjectKind().GroupVersionKind())
-	if !ok {
-		// There's no subscriber for this object kind - nothing to do.
-		return
+	// Publish the event to all subscribers.
+	for _, ch := range q.getSubscriptionsForKind(obj.GetObjectKind().GroupVersionKind()) {
+		ch <- event.GenericEvent{Object: obj}
 	}
-	ch <- event.GenericEvent{Object: obj}
 }
 
 // Subscribe provides a consumer channel where generic events for the provided
 // object kind will be published when the status for the object needs to be
 // updated.
 //
-// Note that more than one consumer can subscribe to a channel for a particular
-// object kind, but that this only represents a single channel: events will not
-// be duplicated and each subscriber will receive events on a first come first
-// serve basis.
+// Please note every call to Subscribe will create a new subscription channel.
 func (q *Queue) Subscribe(gvk schema.GroupVersionKind) chan event.GenericEvent {
-	return q.getOrCreateSubscriptionForKind(gvk)
+	return q.createSubscriptionForKind(gvk)
 }
 
-// getOrCreateSubscriptionForKind returns the subscription channel for the provided object GVK.
-// If the channel does not exist, it will be created.
-func (q *Queue) getOrCreateSubscriptionForKind(gvk schema.GroupVersionKind) chan event.GenericEvent {
+// createSubscriptionForKind creates a new subscription channel for the provided object GVK and returns it.
+func (q *Queue) createSubscriptionForKind(gvk schema.GroupVersionKind) chan event.GenericEvent {
 	q.lock.Lock()
 	defer q.lock.Unlock()
-	ch, ok := q.subscriptions[gvk.String()]
-	if !ok {
-		// If there's no channel built for this kind yet, make it.
-		ch = make(chan event.GenericEvent, q.subscriptionBufferSize)
-		q.subscriptions[gvk.String()] = ch
-	}
-	return ch
+	newSubscriptionCh := make(chan event.GenericEvent, q.subscriptionBufferSize)
```

Comment:
Nit: since the buffer size is only read after the `Queue` is constructed you could put this line above the `q.lock` being locked.

Reply:
True, good point 👍